Comments docs improvements (#1260)
liveblocks/liveblocks: guides/pages/how-to-add-users-to-liveblocks-comments.mdx

## Authenticate your application

The first step is to find an
[authentication guide for your framework](/docs/rooms/authentication) and
authenticate your app, as this is necessary for Comments.

Make sure to follow the metadata step in the guide, and attach the name of your
user, along with the URL of their avatar, as these properties will both be used
in the [default components](/docs/products/comments#Components). Here’s an
example using access token authentication, with an email address as a user’s ID.

```ts title="Metadata in access tokens"
const session = liveblocks.prepareSession("marc@example.com", {
  userInfo: {
    name: "Marc",
    avatar: "https://example.com/marc.png",

    // Your custom metadata
    // ...
  },
});
```

If you’re using ID token authentication, it’ll look a little different.

```ts title="Metadata in ID tokens" isCollapsed isCollapsable
const { status, body } = await liveblocks.identifyUser(
  {
    userId: "marc@example.com",
    groupIds: [], // Optional
  },
  {
    userInfo: {
      name: "Marc",
      avatar: "https://example.com/marc.png",

      // Your custom metadata
      // ...
    },
  }
);
```

> Don’t forget to modify your `UserMetadata` type in `liveblocks.config.ts` to

Reviewer: I just looked and it's already like that in existing authentication guides but shouldn't we use the same name as the type (`UserMeta`) instead of `UserMetadata`, it might be confusing to start using two different names in multiple places.

Author: Good spot, I thought I was using the same name everywhere!

match the metadata format, adding type hints to your editor.

```ts file="liveblocks.config.ts"
type UserMetadata = {
  id: string;

  info: {
    name: string;
    avatar: string;

    // Your custom metadata
    // ...
  };
};
```

> Any metadata you add is made saved and accessible in your Comments app with the

Reviewer: ```suggestion
Any metadata you add is saved and made accessible in your Comments app with the
```

Reviewer: Also, I find this paragraph a bit confusing: users haven't learned about `resolveUsers` yet and they just saw how to declare metadata during authentication, yet here it says this data is saved and accessible via `useUser`, I think it's blurring the two concepts together.

Author: Agree, I'll remove the line!

[`useUser`](/docs/api-reference/liveblocks-react#useUser) hook.

## Resolving users

To show each user’s name and avatar in threads and comments, we need to use
[`resolveUsers`](https://liveblocks.io/docs/api-reference/liveblocks-react#resolveUsers).

<Figure>
  <Image
    src="/assets/tutorials/comments/thread.png"
    alt="Thread with resolved users"
    width={1291}
    height={892}
  />
</Figure>

<Steps>
  <Step>
    <StepTitle>Add the function to your config file</StepTitle>
    <StepContent>
      The `resolveUsers` function is passed as an option to
      [`createRoomContext`](/docs/api-reference/liveblocks-react#createRoomContext)
      at the bottom of your `liveblocks.config.ts` file—let’s add it. This function

>       provides you with `userIds`, an array of user IDs that have created a thread or comment.

Reviewer: ```suggestion
      provides you with `userIds`, an array of user IDs that are present in the current room (e.g. they have created a thread, a comment, added a reaction, etc.).
```

Reviewer: I don't like `present` because that's obviously clashing with Presence but I think we don't want to convey a too strict idea of why user IDs would come up here, they're appear as soon a user is visible so that includes many things, even things like mention suggestions, etc.

      These `userIds` match the IDs set when authenticating users in your app.
      
      ```ts highlight="7-13" file="liveblocks.config.ts"
      // ...
      
      const {
        RoomProvider,
        // ...
      } = createRoomContext(client, {
        async resolveUsers({ userIds }) {
          // ["marc@example.com", ...]
          console.log(userIds);
          
          // Return a list of users
          // ...
        },
      });
      ```

    </StepContent>

  </Step>
  <Step>
    <StepTitle>Return your users</StepTitle>
    <StepContent>
      `resolveUsers` requires you to return a list of users in

>       the `ThreadMetadata["info"]` format we set earlier. Remember

Reviewer: ```suggestion
      the `UserMetadata["info"]` format we set earlier. Remember
```
